src/core/mantle/evaluator/simple/utils.ts

```diff
@@ -0,0 +1,92 @@
+import { JSONPath } from "jsonpath-plus";
+import { pick } from "lodash-es";
+
+import { StyleExpression } from "../../types";
+
+export function getCacheableProperties(styleExpression: StyleExpression, feature?: any) {
+  const properties = pick(feature, getCombinedReferences(styleExpression, feature));
+  return properties;
+}
+
+export function getCombinedReferences(expression: StyleExpression, feature?: any): string[] {
+  if (typeof expression === "string") {
+    return getReferences(expression, feature);
+  } else {
+    const references: string[] = [];
+    for (const [condition, value] of expression.conditions) {
+      references.push(...getReferences(condition, feature), ...getReferences(value, feature));
+    }
+    return references;
+  }
+}
+
+export function getReferences(expression: string, feature?: any): string[] {
+  const result: string[] = [];
+  let exp = expression;
+  let i = exp.indexOf("${");
+  const varExpRegex = /^\$./;
+  const jsonPathCache: Record<string, any[]> = {};
+
+  while (i >= 0) {
+    const openSingleQuote = exp.indexOf("'", i);
+    const openDoubleQuote = exp.indexOf('"', i);
+
+    if (openSingleQuote >= 0 && openSingleQuote < i) {
+      const closeQuote = exp.indexOf("'", openSingleQuote + 1);
+      result.push(exp.substring(0, closeQuote + 1));
+      exp = exp.substring(closeQuote + 1);
+    } else if (openDoubleQuote >= 0 && openDoubleQuote < i) {
+      const closeQuote = exp.indexOf('"', openDoubleQuote + 1);
+      result.push(exp.substring(0, closeQuote + 1));
+      exp = exp.substring(closeQuote + 1);
+    } else {
+      const j = exp.indexOf("}", i);
+
+      if (j < 0) {
+        console.log("Unmatched {.");
+      }
+      const varExp = exp.slice(i + 2, j);
+      if (varExpRegex.test(varExp)) {
+        let res = jsonPathCache[varExp];
+        if (!res) {
+          try {
+            res = JSONPath({ json: feature, path: varExp });
+            jsonPathCache[varExp] = res;
+          } catch (error) {
+            console.log("Invalid JSONPath");
+          }
+        }
+        if (res.length !== 0) {
+          console.log("JSONPathEval: ", res[0]);
+          const keyPath = getObjectPathByValue(feature, res[0]);
```

Comment:
I think JSONPath returns a only evaluated primitive value or object.
So if the result value is same with other field value, does it work?

For example, I assume that feature has the following property.
```
{
  count: 1,
  number: 1,
}
```

And if we write a condition like `["${$.number} === 1", "color('red')"]` then it returns `1`, but keyPath may be `[count]`, right?
And if we change the condition to `["${$.count} === 1", "color('blue')"]` and feature is same, the result will be cached value. Because JSONPath returns same `1` and `keyPath` will be `[count]`.